test/e2e/activator_test.go

```diff
@@ -0,0 +1,176 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"testing"
+	"time"
+	"sync/atomic"
+	"net/http"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/pkg/test/spoof"
+	"github.com/knative/serving/test"
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	pkgTest "github.com/knative/pkg/test"
+)
+
+// TestActivatorOverload makes sure that activator can handle the load when scaling from 0.
+// We need to add a similar test for the User pod overload once the second part of overload handling is done.
+func TestActivatorOverload(t *testing.T) {
+	var (
+		logger  = logging.GetContextLogger(t.Name())
+		clients = Setup(t)
+	)
+	const (
+		// The number of concurrent requests to hit the activator with.
+		// 1000 = the number concurrent connections in Istio.
+		concurrency = 1000
+		// Timeout to wait for the responses.
+		// Ideally we must wait ~30 seconds, TODO: need to figure out where the delta comes from.
+		timeout = 65 * time.Second
+		// How long the service will process the request in ms.
+		serviceSleep = 300
+	)
+
+	names := test.ResourceNames{
+		Service: test.AppendRandomString(configName),
+		Image:   "observed-concurrency",
+	}
+
+	configOptions := test.Options{
+		ContainerConcurrency: 1,
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = map[string]string{"autoscaling.knative.dev/maxScale": "10"}
```

Comment:
I don't understand how we couple possible sustain 1000 concurrent connections (above) with a `containerConcurrency` of 1 and a `maxScale` of 10...

Reply:
Not all of them executed concurrently indeed. Since the requests are buffered in the activator and we have a relatively lengthy timeout for them to finish.